### PR TITLE
bgzf writer

### PR DESCRIPTION
**Pull request by dasnellings**
Dependent on my bgzf reader PR.

Implements writing bgzf files using an underlying gzip writer.

### CODE REVIEW

File: bgzf/writer.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// NewWriter creates a bgzf writer from any input writer.
+func NewWriter(w io.Writer) Writer {
+	var err error
+	var zw Writer
+	zw.w = w
+	zw.zipBlock = NewBlock()
+	zw.compressor = gzip.NewWriter(zw.zipBlock)
+	_, err = zw.compressor.Write([]byte{}) // write header
```
Comment (craiglowe) [inline bgzf/writer.go:30]: I don't totally understand writing the gzip header here.

File: bgzf/writer.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		log.Panic(err)
+	}
+	zw.zipBlock.Reset()
```
Comment (craiglowe) [inline bgzf/writer.go:34]: I see the header write and the reset both here when initializing the writer and then again on Write commands.  Do we need them here?

File: bgzf/writer.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// magicEOF generates the magic EOF byte slice from the hex const.
+func magicEOF() []byte {
+	// empty gzip block per sam specs
+	answer, err := hex.DecodeString(magicHexEOF)
+	if err != nil {
+		log.Panic(err)
+	}
+	return answer
```
Comment (craiglowe) [inline bgzf/writer.go:115]: Should we just store this return value as the const at the top of the file?

**Review by craiglowe — APPROVED** [review]
Great work!  I included a couple questions/comments.

File: bgzf/writer.go (hunk 2 above)
Comment (dasnellings) [reply bgzf/writer.go]: I think I added this here before I trashed the header on every write. This is no longer needed now that it is in the Write function and has been removed.

File: bgzf/writer.go (hunk 1 above)
Comment (dasnellings) [reply bgzf/writer.go]: I think I added this here before I trashed the header on every write. This is no longer needed now that it is in the Write function and has been removed.

File: bgzf/writer.go (hunk 3 above)
Comment (dasnellings) [reply bgzf/writer.go:110]: Good idea. I've made it a private var.

**Review by craiglowe — APPROVED** [review]
Wonderful!

**Review by RavenYL — APPROVED** [review]

File: bgzf/writer.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// writeHeader for custom bgzf header.
+func (w Writer) writeHeader(compSize int) {
+	var header [18]byte
+	header[0] = 31  // gzip ID 1
+	header[1] = 139 // gzip ID 2
+	header[2] = 8   // gzip compression method
+	header[3] = 4   // gzip flags
+	// header[4:8] = mod time uint32
```
Comment (cshoben) [inline bgzf/writer.go:85]: Does this mean that headers 4 through 8 are blank or skipped? I'm trying ti visualize this in my head.

**Review by cshoben — APPROVED** [review]
I appreciate these being two separate PR. Just a conceptual question :)

File: bgzf/writer.go (hunk 4 above)
Comment (dasnellings) [reply bgzf/writer.go:85]: This means that 4-8 are set to 0 (the default value). I initialize the array to have a specific size where everything is set to zero by default. If I want any values to stay zero, I can just ignore them. I just leave the comment here for clarity.

**Pull request merged**